User Profile - Update APIs to work with domain

ywangd opened the pull request:
This PR updates the three existing profile APIs to work with the new
security domain concept:

* Domain info gets recoreded when activating profile
* Domain info is returned appropriately based on the context
* "Same" profile is returned based on checking realm name and domain
  definition

ywangd committed:
User Profile - Update APIs to work with domain

This PR updates the three existing profile APIs to work with the new
security domain concept:
* Domain info gets recoreded when activating profile
* Domain info is returned appropriately based on the context
* "Same" profile is returned based on checking realm name and domain
  definition

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

elasticsearchmachine commented:
Hi @ywangd, I've created a changelog YAML for you.

ywangd committed:
Update docs/changelog/83570.yaml

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileDocument.java:
```diff
@@ -74,8 +67,8 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
 
-        public Profile.ProfileUser toProfileUser(@Nullable String realmDomain) {
-            return new Profile.ProfileUser(username, roles, realm.getName(), realmDomain, email, fullName, displayName, active);
+        public Profile.ProfileUser toProfileUser(@Nullable String domainName) {
```
`domainName` can come from `realm.getDomain().name()`, no? Is there a reason we're not using that?

ywangd replied:
Short answer: We should just use the domain information from the ProfileDocument itself and I updated accordingly.

Long answer: Yes there was a reason. The domain recorded in the document is a snapshot of the configuration at the moment of creation. We agreed that the actual domain information will still be derived from the incoming authentication (if there is one). That is, the final domain info may not be the same as the one recorded in the doc. 
That said, the use cases so far in this PR do not see any difference in the two because the document is either newly created or just updated with the incomig authentication. So the two should always have the identical domain info. For simplicity, I agree that this method should not take parameter and just uses the doc's domain info.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:
```diff
@@ -183,16 +187,31 @@ private void getVersionedDocument(String uid, ActionListener<VersionedDocument>
+                    .getDomain()
+                    .realms()
+                    .stream()
+                    .map(RealmConfig.RealmIdentifier::getName)
```
I thought we were going to rely on the realm name-type tuple for these checks, no?
The fact that realm name is unique in a given node yml config file is not good enough to generally qualify a unique user, as I recall.

ywangd replied:
Yes you are right. I had a different thought (along the direction that we may want ultimately rely only on name). But even that is a thing in the future and for now we should still use both name and type. I updated accordingly.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:
```diff
@@ -400,14 +425,16 @@ private ProfileDocument updateWithSubject(ProfileDocument doc, Subject subject)
+         * (2) replace the domain with the most appropriate one. This means the latest authentication object
+         *     if one exists or extraction from the document itself.
+         */
+        Profile toProfile(Set<String> dataKeys, @Nullable RealmDomain realmDomain) {
```
Same as above, I believe we don't need to supply the domain here, no?

ywangd replied:
See above.

albertzaharovits commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainSingleNodeTests.java:
```diff
@@ -0,0 +1,225 @@
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+public class ProfileDomainSingleNodeTests extends AbstractProfileSingleNodeTestCase {
```
I appreciate the conscientiousness in these tests.

albertzaharovits reviewed (commented):
I had a couple of comments, and I'll take another look afterwards to make sure I comprehend it all.

albertzaharovits commented:
Not part of this PR but do you think we should do another profile search after a new one has been created, in order to fail if there is a race?

ywangd commented:
> Not part of this PR but do you think we should do another profile search after a new one has been created, in order to fail if there is a race?

We definitely need to handle potential racing condition. It should be of its own PR given the likely complexity. I have some idea and we can discuss about it offline.

ywangd committed:
Address feedback

ywangd committed:
Merge remote-tracking branch 'origin/master' into user-profile-update-apis-for-domain

ywangd committed:
fix test

albertzaharovits reviewed (approved):
LGTM
GH is misbehaving for me, so I'll comment in here:
* in tests (IT) it would be nice to cover the activate using the token grant
* the query to search for profiles given a domain, can always use `filter` instead of `should` (later, in a follow-up, maybe we can cache it by domain, to avoid the query building on every activate call)

albertzaharovits commented:
>  always use `filter` instead of `should`

I mean instead of `must`.

ywangd committed:
address feedback

ywangd committed:
Merge remote-tracking branch 'origin/master' into user-profile-update-apis-for-domain

ywangd commented:
> * in tests (IT) it would be nice to cover the activate using the token grant

Good point. I updated the test helper method to always randomly activate profile using each the user password or a token created from it.

> * the query to search for profiles given a domain, can always use `filter` instead of `should` (later, in a follow-up, maybe we can cache it by domain, to avoid the query building on every activate call)

Yes filter is better since we don't care about score. I have replaced all `must` with `filter`.
